src/compiler/checker.ts

```diff
@@ -19944,6 +19945,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return type;
     }
 
+    function getNormalizedTupleType(type: TupleTypeReference, writing: boolean): Type {
+        let normalizedElements: Type[] | undefined;
+        const elements = getTypeArguments(type);
+
+        for (let i = 0; i < elements.length; i++) {
+            if (elements[i].flags & TypeFlags.Simplifiable) {
+                normalizedElements ??= elements.slice(0, i);
+                normalizedElements.push(getSimplifiedType(elements[i], writing));
+            }
+            else {
+                normalizedElements?.push(elements[i]);
+            }
+        }

```

Comment:
Can the logic here use the `sameMap` helper, which has the added benefit that no new type is constructed when `getSimplifiedType` ends up returning the same type?

Reply:
Pushed out 2 commits:
- first one avoids initializing `normalizedElements` if `getSimplifiedType` returns the same type
- second one reused `sameMap` (good tip! I missed it previously)

I think it makes sense to reuse `sameMap` since it does exactly what I want, with the lazy init of the resulting array and all.